File: backend/seiche/market_workbench.py
"""Bounded public research over cached FX and accepted Palimpsest evidence.

No request starts collection, fits an engine, or grants publication rights.
Crosses use the intersection of observation dates, never a filled chart frame.
"""

from __future__ import annotations

import math
import re
import sqlite3
from collections.abc import Mapping
from datetime import UTC, date, datetime, timedelta
from typing import Any

import pandas as pd

from seiche import context_views, methodology, store
from seiche.config import ALL_SERIES, ECB_FX_CURRENCIES
from seiche.palimpsest_china_intake import PalimpsestChinaEconomicContext
from seiche.sources.base import Series

SCHEMA = "seiche.market-workbench.v1"
# The H.10 starred quotes are USD per foreign unit. Other quotes are foreign
# units per USD. EUR's legacy mnemonic is intentionally retained.
FX_MNEMONICS = {
    c: c
    for c in (
        "GBP",
        "JPY",
        "AUD",
        "CAD",
        "CHF",
        "NZD",
        "CNY",
        "INR",
        "KRW",
        "MXN",
        "BRL",
        "ZAR",
        "DKK",
        "HKD",
        "MYR",
        "NOK",
        "SEK",
        "SGD",
        "TWD",
        "THB",
        "LKR",
    )
} | {"EUR": "EURUSD"}
INVERTED = frozenset({"EUR", "GBP", "AUD", "NZD"})
ECB_MNEMONICS = {c: f"ECBFX_{c}" for c in ECB_FX_CURRENCIES}
PROVIDER_CURRENCIES = {
    "h10": tuple(sorted({"USD", *FX_MNEMONICS})),
    "ecb": tuple(sorted({"EUR", *ECB_MNEMONICS})),
}
CURRENCIES = tuple(
    sorted(set(PROVIDER_CURRENCIES["h10"]) | set(PROVIDER_CURRENCIES["ecb"]))
)
INPUT_SCHEMA = {
    "type": "object",
    "additionalProperties": False,
    "properties": {
        "provider": {"type": "string", "enum": ["h10", "ecb"], "default": "h10"},
        "base": {"type": "string", "enum": list(CURRENCIES), "default": "USD"},
        "quote": {"type": "string", "enum": list(CURRENCIES), "default": "CNY"},
        "days": {"type": "integer", "minimum": 30, "maximum": 3650, "default": 365},
        "china_series": {"type": "string", "maxLength": 120},
    },
}


def selection(args: object) -> dict[str, Any]:
    if not isinstance(args, dict) or set(args) - set(INPUT_SCHEMA["properties"]):
        raise ValueError(
            "expected provider, base, quote, days and optional china_series"
        )
    out = {
        "provider": "h10",
        "base": "USD",
        "quote": "CNY",
        "days": 365,
        "china_series": "",
        **args,
    }
    if type(out["provider"]) is not str or out["provider"] not in PROVIDER_CURRENCIES:
        raise ValueError("provider must be h10 or ecb")
    for name in ("base", "quote"):
        if (
            not isinstance(out[name], str)
            or out[name] not in PROVIDER_CURRENCIES[out["provider"]]
        ):
            raise ValueError(f"{name} must be a supported currency code")
    if out["base"] == out["quote"]:
        raise ValueError("base and quote must differ")
    if type(out["days"]) is not int or not 30 <= out["days"] <= 3650:
        raise ValueError("days must be an integer between 30 and 3650")
    if not isinstance(out["china_series"], str) or (
        out["china_series"]
        and not re.fullmatch(r"cn\.wdi\.[a-z0-9_]{1,110}", out["china_series"])
    ):
        raise ValueError("china_series must be a cn.wdi series identifier")
    return out


def _clock(value: str) -> datetime | None:
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        return parsed.astimezone(UTC) if parsed.tzinfo is not None else None
    except (ValueError, TypeError, AttributeError):
        return None


def _normalized(
    currency: str,
    series: Series | None,
    now: datetime,
    days: int,
    provider: str = "h10",
) -> tuple[dict[str, float], dict | None]:
    if series is None:
        return {}, None
    mnemonic = (FX_MNEMONICS if provider == "h10" else ECB_MNEMONICS)[currency]
    spec = ALL_SERIES[mnemonic]
    fetched_at = _clock(series.fetched_at)
    # Cache metadata must still match the public, reviewed H.10 registry.
    if (
        series.mnemonic != mnemonic
        or series.source != spec.source
        or series.remote_id != spec.remote_id
        or series.unit != spec.unit
        or series.freq != spec.freq
        or methodology.csv_restriction(mnemonic)
        or fetched_at is None
        or fetched_at > now
    ):
        return {}, None
    cutoff = now.date() - timedelta(days=days)
    values: dict[str, float] = {}
    for timestamp, value in series.points.items():
        try:
            day = pd.Timestamp(timestamp).date()
            number = float(value)
        except (ValueError, TypeError, OverflowError):
            continue
        if (
            not cutoff <= day <= min(now.date(), fetched_at.date())
            or not math.isfinite(number)
            or number <= 0
        ):
            continue
        key = day.isoformat()
        # An ambiguous date is not a public reference observation.
        if key in values:
            return {}, None
        normalized = (
            1.0 / number if provider == "h10" and currency in INVERTED else number
        )
        if math.isfinite(normalized) and normalized > 0:
            values[key] = normalized
    provenance = {
        "mnemonic": mnemonic,
        "source_id": spec.remote_id,
        "publisher": "Board of Governors of the Federal Reserve System",
        "transport": "Federal Reserve Bank of St. Louis FRED",
        "source_url": f"https://fred.stlouisfed.org/series/{spec.remote_id}",
        "publisher_url": "https://www.federalreserve.gov/releases/h10/current/",
        "raw_unit": spec.unit,
        "raw_quote_convention": "USD per currency"
        if currency in INVERTED
        else "currency per USD",
        "fetched_at": series.fetched_at,
        "source_publication_time": None,
        "redistribution_status": "allowed",
        "vintage": "current_cached_vintage",
    }
    if provider == "ecb":
        provenance.update(
            publisher="European Central Bank",
            transport="ECB euro foreign exchange reference XML",
            source_url="https://www.ecb.europa.eu/stats/eurofxref/eurofxref-hist.xml",
            publisher_url="https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html",
            raw_quote_convention="currency per EUR",
            attribution="Source: European Central Bank. ECB reference data is freely available from the ECB; calculated crosses and statistics are Seiche transformations.",
            terms_url="https://www.ecb.europa.eu/services/using-our-site/disclaimer/html/index.en.html",
        )
    return dict(sorted(values.items())), provenance


def _pair(
    base: str,
    quote: str,
    data: Mapping[str, dict[str, float]],
    sources: Mapping[str, dict | None],
    now: datetime,
    provider: str = "h10",
) -> tuple[dict, list[dict]]:
    anchor = "USD" if provider == "h10" else "EUR"
    legs = [currency for currency in (base, quote) if currency != anchor]
    dates = set(data.get(legs[0], {}))
    for leg in legs[1:]:
        dates.intersection_update(data.get(leg, {}))
    history = []
    for day in sorted(dates):
        numerator = 1.0 if quote == anchor else data[quote][day]
        denominator = 1.0 if base == anchor else data[base][day]
        value = numerator / denominator
        if math.isfinite(value) and value > 0:
            history.append({"date": day, "value": value})
    latest = history[-1] if history else None
    age = (now.date() - date.fromisoformat(latest["date"])).days if latest else None
    grace = 7 if provider == "h10" else 3
    status = (
        "unavailable"
        if age is None
        else "fresh"
        if age <= grace
        else "aging"
        if age <= grace * 2
        else "stale"
        if age <= grace * 6
        else "dead"
    )
    observed = (base == "USD" and quote not in INVERTED) or (
        quote == "USD" and base in INVERTED
    )
    if provider == "ecb":
        observed = base == "EUR"

    def change(periods: int) -> float | None:
        value = (
            (history[-1]["value"] / history[-periods - 1]["value"] - 1) * 100
            if len(history) > periods
            else None
        )
        return value if value is not None and math.isfinite(value) else None

    vol = None
    # A long gap must not masquerade as a single daily return.
    if len(history) >= 21:
        sample = history[-21:]
        if all(
            (date.fromisoformat(b["date"]) - date.fromisoformat(a["date"])).days <= 7
            for a, b in zip(sample, sample[1:])
        ):
            returns = [
                math.log(b["value"]) - math.log(a["value"])
                for a, b in zip(sample, sample[1:])
            ]
            mean = sum(returns) / len(returns)
            vol = (
                math.sqrt(sum((r - mean) ** 2 for r in returns) / len(returns) * 252)
                * 100
            )
    return {
        "pair": f"{base}/{quote}",
        "base_currency": base,
        "quote_currency": quote,
        "provider": provider,
        "value": latest["value"] if latest else None,
        "as_of": latest["date"] if latest else None,
        "status": status,
        "evidence_status": ("observed" if observed else "derived")
        if latest
        else "unavailable",
        "unit": f"{quote} per {base}",
        "observation_count": len(history),
        "age_days": age,
        **{f"change_{n}obs_pct": change(n) for n in (1, 5, 20, 60)},
        "realized_vol_20obs_pct": vol,
        "sources": [sources[c] for c in legs if sources.get(c)],
        "reason": None
        if latest
        else "No eligible matching-date observations for the selected window.",
    }, history


CHINA_CHANNELS = (
    (
        "funding",
        "Domestic money and credit",
        "Money growth, private credit and inflation describe the domestic funding backdrop; annual values cannot establish today's repo pressure.",
        ("broad_money", "bank_credit", "inflation"),
    ),
    (
        "external",
        "External balance and yuan demand",
        "Current-account, trade and investment evidence describes potential foreign-currency supply and demand; it does not measure today's spot order flow.",
        ("current_account", "exports", "imports", "trade", "fdi"),
    ),
    (
        "reserves",
        "Reserves and external buffers",
        "Reserve levels and import cover describe external buffers; a stock change alone cannot establish intervention.",
        ("reserves", "external_debt"),
    ),
    (
        "activity",
        "Growth and policy backdrop",
        "Growth and investment place monetary and currency policy in context. Source vintages and annual frequency remain visible.",
        ("gdp", "investment", "industry", "manufacturing"),
    ),
)


def _china(context: object, selected: str, usd_cny: dict) -> dict:
    out: dict[str, Any] = {
        "status": "unavailable",
        "reason": "No owner-accepted Palimpsest economic export is configured.",
        "economic_context": {},
        "series": [],
        "selected_series": selected or None,
        "history": [],
        "fx": usd_cny,
        "channels": [],
        "gaps": [
            {
                "id": "cfets",
                "label": "DR007 / FDR007 / SHIBOR",
                "status": "restricted",
                "reason": "Publisher rights are required for public benchmark values. FDR007 and DR007 are distinct series.",
            },
            {
                "id": "cnh",
                "label": "Offshore CNH and CNY–CNH basis",
                "status": "unavailable",
                "reason": "No entitled, matching-time offshore quote is available; a CNY reference is not CNH.",
            },
            {
                "id": "forwards",
                "label": "FX forwards, swaps and NDFs",
                "status": "unavailable",
                "reason": "No entitled tenor curves or executable dealer quotes are configured.",
            },
            {
                "id": "fixing",
                "label": "PBOC central parity deviation",
                "status": "unavailable",
                "reason": "The selected CNY reference is not the PBOC central parity fixing.",
            },
        ],
    }
    if isinstance(context, PalimpsestChinaEconomicContext) and context.owner_attested:
        economic = context.to_dict()
        if economic.get("rights", {}).get("decision") == "allowed":
            economic["publication_status"] = "provisional"
            rows = [
                row.public_record(accepted_at=context.accepted_at)
                for row in context.current_observations
            ]
            for row in rows:
                row["label"] = (
                    row["series_id"]
                    .removeprefix("cn.wdi.")
                    .replace("_", " ")
                    .capitalize()
                )
            selected = selected or (rows[0]["series_id"] if rows else "")
            if selected and selected not in {row["series_id"] for row in rows}:
                raise ValueError("china_series is not present in the accepted export")
            by_period = {}

            def rank(row: dict) -> tuple:
                return (
                    row["revision"],
                    row["released_at"],
                    row["collected_at"],
                    row["observation_id"],
                )

            for observation in context.observations:
                if observation.series_id != selected:
                    continue
                row = observation.public_record(accepted_at=context.accepted_at)
                key = row["period_end"]
                if key not in by_period or rank(row) > rank(by_period[key]):
                    by_period[key] = row
            history = [by_period[key] for key in sorted(by_period)]
            out.update(
                status="structural",
                reason=None,
                economic_context=economic,
                series=rows,
                selected_series=selected or None,
                history=history[-100:],
                history_total=len(history),
                history_truncated=len(history) > 100,
                history_vintage="latest_revision_in_accepted_export",
            )
    for identifier, title, reading, tokens in CHINA_CHANNELS:
        ids = [
            row["series_id"]
            for row in out["series"]
            if any(t in row["series_id"] for t in tokens)
        ]
        out["channels"].append(
            {
                "id": identifier,
                "title": title,
                "reading": reading,
                "series_ids": ids,
                "available_series": len(ids),
            }
        )
    return out


def project(
    series: Mapping[str, Series],
    args: object = None,
    *,
    china_context: object = None,
    evaluated_at: datetime | None = None,
) -> dict:
    query = selection({} if args is None else args)
    now = evaluated_at or datetime.now(UTC)
    if now.tzinfo is None:
        raise ValueError("evaluation clock must be timezone-aware")
    now = now.astimezone(UTC)
    provider = query["provider"]
    mnemonics = FX_MNEMONICS if provider == "h10" else ECB_MNEMONICS
    data, sources = {}, {}
    for currency, mnemonic in mnemonics.items():
        data[currency], sources[currency] = _normalized(
            currency, series.get(mnemonic), now, query["days"], provider
        )
    rows = []
    selected_history = []
    for quote in PROVIDER_CURRENCIES[provider]:
        if quote == query["base"]:
            continue
        row, history = _pair(query["base"], quote, data, sources, now, provider)
        rows.append(row)
        if quote == query["quote"]:
            selected_history = history
    usd_cny, _ = _pair("USD", "CNY", data, sources, now, provider)
    china = _china(china_context, query["china_series"], usd_cny)
    available = sum(row["value"] is not None for row in rows)
    return {
        "schema": SCHEMA,
        "generated_at": now.isoformat(),
        "context_only": True,
        "status": "available"
        if available == len(rows) and china["series"]
        else "partial"
        if available or china["series"]
        else "unavailable",
        "selection": query,
        "forex": {
            "base_currency": query["base"],
            "quote_currency": query["quote"],
            "quote_convention": "quote currency units per one base currency",
            "reference_only": True,
            "provider": provider,
            "currencies": list(PROVIDER_CURRENCIES[provider]),
            "rows": rows,
            "history": selected_history,
            "coverage": {
                "declared_pairs": len(rows),
                "available_pairs": available,
                "missing_pairs": len(rows) - available,
                "fresh_pairs": sum(r["status"] == "fresh" for r in rows),
                "returned_observations": len(selected_history),
            },
            "methodology": [
                (
                    "Federal Reserve H.10 daily reference observations delivered through FRED."
                    if provider == "h10"
                    else "Source: European Central Bank. ECB reference data is freely available from the ECB; calculated crosses and statistics are Seiche transformations."
                )
                + " Not executable quotes, intraday data, or forward rates.",
                f"Crosses divide quote-per-{'USD' if provider == 'h10' else 'EUR'} by base-per-{'USD' if provider == 'h10' else 'EUR'} on exact shared observation dates; no forward filling or mixed-date quote construction.",
                "Changes use 1, 5, 20 and 60 observed intervals. Positive means the quote currency weakened against the base currency.",
                "Volatility uses 20 log returns, population standard deviation and sqrt(252) annualization; any interval longer than seven days makes it unavailable.",
                (
                    "Age states use seven calendar days for fresh, fourteen for aging, forty-two for stale, then dead, allowing for H.10 weekly delivery."
                    if provider == "h10"
                    else "Age states use three calendar days for fresh, six for aging, eighteen for stale, then dead for ECB daily reference data."
                )
                + " These are reference-data age labels, not market-open checks.",
                "History is the current cached vintage, not a point-in-time backtest. Fetch and response clocks do not advance observation dates.",
            ],
        },
        "china": china,
        "money_markets": {
            "catalog_url": "/api/v2/markets",
            "series_url_template": "/api/v2/markets/{market_id}/series",
        },
        "eligibility": {"scoring": False, "forecast": False, "execution": False},
    }


def read(args: object = None) -> dict:
    query = selection({} if args is None else args)
    now = datetime.now(UTC)
    mnemonics = FX_MNEMONICS if query["provider"] == "h10" else ECB_MNEMONICS
    captures = {}
    failure = None
    window = {
        "start": (now.date() - timedelta(days=query["days"])).isoformat(),
        "end": now.date().isoformat(),
    }
    try:
        if query["provider"] == "ecb":
            cached, captures = store.load_series_window_snapshot(
                tuple(mnemonics.values()),
                **window,
                blob_keys=("ecb_fx:latest", "ecb_fx:full-history"),
            )
            latest = _public_capture(captures.get("ecb_fx:latest"), now)
            if latest is None:
                cached = {}
                failure = "ECB capture metadata is unavailable or invalid."
            else:
                cached = {
                    key: value
                    for key, value in cached.items()
                    if _clock(value.fetched_at) == _clock(latest["fetched_at"])
                    and value.asof == latest["last_observation_date"]
                }
        else:
            cached = store.load_series_window(tuple(mnemonics.values()), **window)
    except (sqlite3.Error, ValueError):
        cached = {}
        failure = "The cached source window could not be verified."
    # Invalid configured acceptance/rights raises; it must never masquerade as
    # an accepted empty or refreshed economic panel.
    china = context_views.public_china_economic_context()
    result = project(cached, query, china_context=china, evaluated_at=now)
    if failure:
        result["forex"]["reason"] = failure
    if query["provider"] == "ecb":
        latest = _public_capture(captures.get("ecb_fx:latest"), now)
        full = _public_capture(captures.get("ecb_fx:full-history"), now)
        result["forex"]["capture_documents"] = {
            "latest": latest,
            "full_history": full,
            "boundary": "Each hash covers its named downloaded document and date interval, not the entire merged history. Retired currencies in full-history captures are excluded from current coverage.",
        }
        if latest:
            for row in [*result["forex"]["rows"], result["china"]["fx"]]:
                for source in row["sources"]:
                    source["source_url"] = latest["source_url"]
                    source["latest_capture_sha256"] = latest["evidence_sha256"]
                    source["capture_interval"] = [
                        latest["first_observation_date"],
                        latest["last_observation_date"],
                    ]
    return result


def _public_capture(value: object, now: datetime) -> dict | None:
    if not isinstance(value, dict):
        return None
    suffix = {"daily": "daily", "history90d": "hist-90d", "history": "hist"}.get(
        str(value.get("mode"))
    )
    clock = _clock(value.get("fetched_at"))
    if (
        value.get("schema") != "seiche.ecb-fx-capture.v1"
        or value.get("source") != "ecb_fx"
        or value.get("base_currency") != "EUR"
        or suffix is None
        or value.get("source_url")
        != f"https://www.ecb.europa.eu/stats/eurofxref/eurofxref-{suffix}.xml"
        or clock is None
        or clock > now
        or not isinstance(value.get("evidence_sha256"), str)
        or re.fullmatch(r"[0-9a-f]{64}", value["evidence_sha256"]) is None
    ):
        return None
    try:
        first = date.fromisoformat(value["first_observation_date"])
        last = date.fromisoformat(value["last_observation_date"])
        if not date(1999, 1, 1) <= first <= last <= clock.date():
            return None
    except (KeyError, TypeError, ValueError):
        return None
    return {
        key: value.get(key)
        for key in (
            "schema",
            "source",
            "source_url",
            "evidence_sha256",
            "fetched_at",
            "source_publication_time",
            "byte_count",
            "first_observation_date",
            "last_observation_date",
            "observation_count",
            "mode",
        )
    }

File: frontend/src/marketWorkbench.ts
/** Display and export contracts for the public market workbench. */
export interface FxSource {
  mnemonic: string;
  source_id: string;
  source_url: string | null;
  raw_unit: string;
  fetched_at: string | null;
  raw: Record<string, unknown>;
}

export interface FxRow {
  pair: string;
  base_currency: string;
  quote_currency: string;
  value: number | null;
  as_of: string | null;
  status: string;
  evidence_status: string;
  unit: string;
  change_1obs_pct: number | null;
  change_5obs_pct: number | null;
  change_20obs_pct: number | null;
  change_60obs_pct: number | null;
  realized_vol_20obs_pct: number | null;
  observation_count: number;
  sources: FxSource[];
  reason: string | null;
}

export interface HistoryPoint { date: string; value: number }
export interface ChinaObservation {
  series_id: string;
  label: string;
  value: number | null;
  unit: string;
  period_end: string;
  period_start: string;
  released_at: string | null;
  collected_at: string | null;
  accepted_at: string | null;
  evidence_url: string | null;
  source_id: string;
  market_channels: string[];
  revision: unknown;
  raw: Record<string, unknown>;
}
export interface ChinaChannel {
  id: string; title: string; reading: string; series_ids: string[]; available_series: number;
}
export interface WorkbenchSelection {
  base: string; quote: string; days: number; provider?: "h10" | "ecb"; china_series?: string | null;
}
export interface MarketWorkbenchData {
  schema: "seiche.market-workbench.v1";
  generated_at: string;
  selection: WorkbenchSelection;
  forex: {
    base_currency: string;
    quote_currency: string;
    quote_convention: string;
    currencies: string[];
    rows: FxRow[];
    history: HistoryPoint[];
    coverage: { declared_pairs: number; available_pairs: number; missing_pairs: number; returned_observations: number };
    methodology: string[];
  };
  china: {
    status: string;
    reason: string | null;
    economic_context: Record<string, unknown> | null;
    series: ChinaObservation[];
    selected_series: string | null;
    history: ChinaObservation[];
    channels: ChinaChannel[];
    gaps: Array<{ id: string; label: string; status: string; reason: string }>;
    fx: FxRow | null;
  };
  raw: Record<string, unknown>;
}

const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;
const CURRENCY = /^[A-Z]{3}$/;
const DECIMAL = /^-?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$/;

function object(value: unknown, name: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`Invalid ${name} response`);
  return value as Record<string, unknown>;
}
function string(value: unknown, fallback = ""): string { return typeof value === "string" ? value : fallback; }
function strings(value: unknown): string[] { return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : []; }
function list(value: unknown, name: string): unknown[] {
  if (!Array.isArray(value)) throw new Error(`Missing ${name} records`);
  return value;
}
export function finiteValue(value: unknown): number | null {
  if (value === null || value === undefined) return null;
  const parsed = typeof value === "number" ? value
    : typeof value === "string" && DECIMAL.test(value) ? Number(value) : NaN;
  if (!Number.isFinite(parsed)) throw new Error("Invalid numeric observation");
  return parsed;
}
function count(value: unknown): number {
  const parsed = finiteValue(value);
  if (parsed === null || !Number.isInteger(parsed) || parsed < 0) throw new Error("Invalid observation count");
  return parsed;
}
function date(value: unknown): string {
  if (typeof value !== "string" || !ISO_DATE.test(value) || !Number.isFinite(Date.parse(value))
    || new Date(value).toISOString().slice(0, 10) !== value) throw new Error("Invalid observation date");
  return value;
}
export function safeEvidenceUrl(value: unknown): string | null {
  if (typeof value !== "string") return null;
  try {
    const url = new URL(value);
    return url.protocol === "https:" && !url.username && !url.password ? url.href : null;
  } catch { return null; }
}
function currency(value: unknown): string {
  if (typeof value !== "string" || !CURRENCY.test(value)) throw new Error("Invalid currency code");
  return value;
}
function fxRow(value: unknown): FxRow {
  const row = object(value, "currency");
  const base = currency(row.base_currency), quote = currency(row.quote_currency);
  const numeric = finiteValue(row.value);
  if (row.pair !== `${base}/${quote}` || row.unit !== `${quote} per ${base}` || base === quote) throw new Error("Currency label and quote convention disagree");
  if (numeric !== null && (numeric <= 0 || !row.as_of)) throw new Error("Invalid dated reference rate");
  return {
    pair: string(row.pair), base_currency: base, quote_currency: quote,
    value: numeric, as_of: row.as_of ? date(row.as_of) : null,
    status: string(row.status, "unavailable"), evidence_status: string(row.evidence_status, "unavailable"),
    unit: string(row.unit), change_1obs_pct: finiteValue(row.change_1obs_pct),
    change_5obs_pct: finiteValue(row.change_5obs_pct), change_20obs_pct: finiteValue(row.change_20obs_pct),
    change_60obs_pct: finiteValue(row.change_60obs_pct), realized_vol_20obs_pct: finiteValue(row.realized_vol_20obs_pct),
    observation_count: count(row.observation_count), reason: string(row.reason) || null,
    sources: list(row.sources, "currency source").map((value) => {
      const source = object(value, "currency source");
      return { mnemonic: string(source.mnemonic), source_id: string(source.source_id), source_url: safeEvidenceUrl(source.source_url), raw_unit: string(source.raw_unit), fetched_at: string(source.fetched_at) || null, raw: source };
    }),
  };
}
function chinaObservation(value: unknown): ChinaObservation {
  const row = object(value, "China observation");
  if (!string(row.series_id)) throw new Error("China observation has no series identity");
  return {
    series_id: string(row.series_id), label: string(row.label) || string(row.series_id).replace(/^cn\.wdi\./, "").replaceAll("_", " "),
    value: finiteValue(row.value), unit: string(row.unit), period_end: date(row.period_end),
    period_start: date(row.period_start), released_at: string(row.released_at) || null,
    collected_at: string(row.collected_at) || null, accepted_at: string(row.accepted_at) || null,
    evidence_url: safeEvidenceUrl(row.evidence_url), source_id: string(row.source_id),
    market_channels: strings(row.market_channels), revision: row.revision, raw: row,
  };
}

export function normalizeWorkbench(value: unknown, expected: WorkbenchSelection): MarketWorkbenchData {
  const raw = object(value, "workbench");
  if (raw.schema !== "seiche.market-workbench.v1") throw new Error("Unrecognized workbench schema");
  const selection = object(raw.selection, "selection");
  if (selection.provider !== (expected.provider ?? "h10")) throw new Error("The response does not match the selected reference-rate provider");
  if (selection.base !== expected.base || selection.quote !== expected.quote || selection.days !== expected.days) {
    throw new Error("The response does not match the selected currencies and window");
  }
  const forex = object(raw.forex, "forex");
  if (forex.base_currency !== expected.base || forex.quote_currency !== expected.quote || forex.reference_only !== true) {
    throw new Error("The response does not establish the selected reference-rate convention");
  }
  const rows = list(forex.rows, "currency").map(fxRow);
  if (rows.some((row) => row.base_currency !== expected.base)
    || new Set(rows.map((row) => row.quote_currency)).size !== rows.length) throw new Error("Ambiguous currency comparison");
  const history = list(forex.history, "FX history").map((value) => {
    const point = object(value, "FX history");
    const numeric = finiteValue(point.value);
    if (numeric === null || numeric <= 0) throw new Error("Invalid FX history value");
    return { date: date(point.date), value: numeric };
  }).sort((a, b) => a.date.localeCompare(b.date));
  if (new Set(history.map((row) => row.date)).size !== history.length) throw new Error("Duplicate FX history date");
  const coverage = object(forex.coverage, "coverage");
  const china = object(raw.china, "China context");
  const series = list(china.series, "China series").map(chinaObservation);
  const chinaHistory = list(china.history, "China history").map(chinaObservation).sort((a, b) => a.period_end.localeCompare(b.period_end));
  const selected = string(china.selected_series) || null;
  if (chinaHistory.some((row) => row.series_id !== selected)
    || new Set(chinaHistory.map((row) => row.period_end)).size !== chinaHistory.length
    || new Set(chinaHistory.map((row) => row.unit)).size > 1) throw new Error("Ambiguous China history series, period or unit");
  if (expected.china_series && selected !== expected.china_series && chinaHistory.length) throw new Error("The response does not match the selected China series");
  const candidate = china.economic_context ? object(china.economic_context, "accepted China context") : null;
  const context = candidate && Object.keys(candidate).length ? candidate : null;
  if ((series.length || chinaHistory.length) && (!context || context.context_only !== true || context.scoring_eligible !== false)) {
    throw new Error("China observations lack their structural evidence boundary");
  }
  return {
    schema: "seiche.market-workbench.v1", generated_at: string(raw.generated_at),
    selection: { base: expected.base, quote: expected.quote, days: expected.days, provider: expected.provider ?? "h10", china_series: string(selection.china_series) || null },
    forex: {
      base_currency: expected.base, quote_currency: expected.quote, quote_convention: string(forex.quote_convention),
      currencies: list(forex.currencies, "currency register").map(currency), rows, history,
      coverage: { declared_pairs: count(coverage.declared_pairs), available_pairs: count(coverage.available_pairs), missing_pairs: count(coverage.missing_pairs), returned_observations: count(coverage.returned_observations) },
      methodology: strings(forex.methodology),
    },
    china: {
      status: string(china.status, "unavailable"), reason: string(china.reason) || null, economic_context: context,
      series, selected_series: selected, history: chinaHistory,
      channels: list(china.channels, "China channel").map((value) => {
        const channel = object(value, "China channel");
        return { id: string(channel.id), title: string(channel.title), reading: string(channel.reading), series_ids: strings(channel.series_ids), available_series: count(channel.available_series) };
      }),
      gaps: list(china.gaps, "China coverage gap").map((value) => {
        const gap = object(value, "China gap");
        return { id: string(gap.id), label: string(gap.label), status: string(gap.status), reason: string(gap.reason) };
      }),
      fx: china.fx ? fxRow(china.fx) : null,
    }, raw,
  };
}

export function filterFxRows(rows: readonly FxRow[], query: string): FxRow[] {
  const tokens = query.trim().toLowerCase().split(/\s+/).filter(Boolean);
  return rows.filter((row) => tokens.every((token) => [row.pair, row.unit, row.quote_currency, row.status, ...row.sources.map((source) => source.source_id)].join(" ").toLowerCase().includes(token)));
}

/** X coordinates use elapsed time: a source holiday must not become an invented observation. */
export function seriesPlot(points: readonly HistoryPoint[], width = 800) {
  const valid = points.filter((point) => Number.isFinite(point.value) && ISO_DATE.test(point.date) && Number.isFinite(Date.parse(point.date)))
    .slice().sort((a, b) => a.date.localeCompare(b.date));
  if (!valid.length) return null;
  const values = valid.map((point) => point.value);
  const low = Math.min(...values), high = Math.max(...values);
  const padding = high === low ? Math.max(Math.abs(low) * .01, .01) : (high - low) * .1;
  const min = low - padding, max = high + padding;
  const start = Date.parse(valid[0].date), end = Date.parse(valid[valid.length - 1].date);
  const plotted = valid.map((point) => ({ ...point,
    x: start === end ? (width + 40) / 2 : 80 + (Date.parse(point.date) - start) / (end - start) * (width - 120),
    y: 22 + (max - point.value) / (max - min) * 200,
  }));
  return { min, max, points: plotted, path: plotted.map((point, i) => `${i ? "L" : "M"}${point.x.toFixed(2)},${point.y.toFixed(2)}`).join(" ") };
}

function csvCell(value: unknown): string {
  if (value === null || value === undefined) return "";
  let text = typeof value === "object" ? JSON.stringify(value) : String(value);
  // Preserve negative numeric values while protecting spreadsheet text cells.
  if (typeof value !== "number" && /^[\s]*[=+@-]/.test(text)) text = `'${text}`;
  return /[",\r\n]/.test(text) ? `"${text.replaceAll('"', '""')}"` : text;
}
function csv(headers: string[], rows: unknown[][]): string {
  return [headers, ...rows].map((row) => row.map(csvCell).join(",")).join("\r\n") + "\r\n";
}
export function fxHistoryCsv(data: MarketWorkbenchData): string {
  const row = data.forex.rows.find((row) => row.quote_currency === data.selection.quote);
  return csv(["date", "value", "base_currency", "quote_currency", "provider", "unit", "evidence_status", "generated_at", "sources", "quote_convention"],
    data.forex.history.map((point) => [point.date, point.value, data.selection.base, data.selection.quote, data.selection.provider, row?.unit, row?.evidence_status, data.generated_at, row?.sources.map((source) => source.raw), data.forex.quote_convention]));
}
export function chinaHistoryCsv(data: MarketWorkbenchData): string {
  return csv(["series_id", "period_start", "period_end", "value", "unit", "released_at", "collected_at", "accepted_at", "source_id", "evidence_url", "revision", "metadata"],
    data.china.history.map((row) => [row.series_id, row.period_start, row.period_end, row.value, row.unit, row.released_at, row.collected_at, row.accepted_at, row.source_id, row.evidence_url, row.revision, row.raw]));
}
